### PR TITLE
Add Visitor Pattern Infrastructure to Painless user/ir Trees

### PR DESCRIPTION
**Pull request by jdconrad**
This change adds the ability to walk either the user and/or the ir tree through a simple visitor pattern with callbacks from the nodes. This allows phases (semantic, write, etc.) to execute externally from the nodes within the trees and allows for another step in greater extensibility moving forward.

The change adds the interfaces UserTreeVisitor and IRTreeVisitor to extend from. Each user tree node and ir tree node support callbacks into these visitors (a mostly mechanical change).

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DeclarationBlockNode.java [hunk 1]
```diff
@@ -21,12 +21,13 @@
...
 import org.elasticsearch.painless.ClassWriter;
 import org.elasticsearch.painless.MethodWriter;
+import org.elasticsearch.painless.phase.IRTreeVisitor;
 import org.elasticsearch.painless.symbol.WriteScope;
 
 import java.util.ArrayList;
 import java.util.List;
 
-public final class DeclarationBlockNode extends StatementNode {
+public class DeclarationBlockNode extends StatementNode {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DeclarationBlockNode.java:30]: why is this not final anymore?

Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/DeclarationBlockNode.java:30]: For consistency and extensibility. None of the nodes are final as they all have the potential to be used as super classes for new nodes. There are couple in the PR that I found going through them that had been overlooked.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ANode.java [hunk 2]
```diff
@@ -55,6 +56,13 @@ public Location getLocation() {
         return location;
     }
 
+    /**
+     * Callback to visit a user tree node.
+     */
+    public <Input, Output> Output visit(UserTreeVisitor<Input, Output> userTreeVisitor, Input input) {
+        throw new UnsupportedOperationException("cannot visit user node type [" + getClass().getCanonicalName() + "]");
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ANode.java:63]: Why is this here just to throw an `UnsupportedOp`?

Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ANode.java:63]: The idea of this is the nodes know how to callback the visitor, but the visitor itself can just call visit on any given node without knowing which specific node it is. Take EBinary as an example, you may want to call leftNode.visit(...) and rightNode.visit(...) without knowing what exactly each node is. The exception here is more of a runtime assert that we tried to visit a node that doesn't support the visitor pattern when it probably should.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/IRTreeVisitor.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+import org.elasticsearch.painless.ir.StaticNode;
+import org.elasticsearch.painless.ir.ThrowNode;
+import org.elasticsearch.painless.ir.TryNode;
+import org.elasticsearch.painless.ir.TypedCaptureReferenceNode;
+import org.elasticsearch.painless.ir.TypedInterfaceReferenceNode;
+import org.elasticsearch.painless.ir.UnaryMathNode;
+import org.elasticsearch.painless.ir.VariableNode;
+import org.elasticsearch.painless.ir.WhileLoopNode;
+
+public interface IRTreeVisitor<Input, Output> {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/IRTreeVisitor.java:81]: If we don't expect implementations to implement _every_ function, then add a no-op default implementation for these.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/UserTreeVisitor.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import org.elasticsearch.painless.node.SFor;
+import org.elasticsearch.painless.node.SFunction;
+import org.elasticsearch.painless.node.SIf;
+import org.elasticsearch.painless.node.SIfElse;
+import org.elasticsearch.painless.node.SReturn;
+import org.elasticsearch.painless.node.SThrow;
+import org.elasticsearch.painless.node.STry;
+import org.elasticsearch.painless.node.SWhile;
+
+public interface UserTreeVisitor<Input, Output> {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/UserTreeVisitor.java:68]: If we don't expect implementations to implement _every_ function, then add a no-op default implementation for these.

**Review by stu-elastic — APPROVED** [review]
You get the :gear: :trophy: for mechanical changes!

Nice easy PR, good design analogous to ANTLR.  A few comments to address before merging.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. Will address your comments soon.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/UserTreeVisitor.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/UserTreeVisitor.java:68]: Good call. Added base versions of these for this use case.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/IRTreeVisitor.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/IRTreeVisitor.java:81]: Good call. Added base versions of these for this use case.
